File: edu/src/main/java/io/agora/edu/classroom/SmallClassActivity.java
package io.agora.edu.classroom;

import android.graphics.Rect;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.tabs.TabLayout;
import com.herewhite.sdk.domain.GlobalState;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.List;
import java.util.Map;

import butterknife.BindView;
import butterknife.OnClick;
import io.agora.edu.R;
import io.agora.education.api.EduCallback;
import io.agora.education.api.base.EduError;
import io.agora.education.api.message.EduChatMsg;
import io.agora.education.api.message.EduMsg;
import io.agora.education.api.room.EduRoom;
import io.agora.education.api.room.data.EduRoomChangeType;
import io.agora.education.api.statistics.ConnectionState;
import io.agora.education.api.statistics.NetworkQuality;
import io.agora.education.api.stream.data.EduStreamEvent;
import io.agora.education.api.stream.data.EduStreamInfo;
import io.agora.education.api.user.EduStudent;
import io.agora.education.api.user.EduUser;
import io.agora.education.api.user.data.EduUserEvent;
import io.agora.education.api.user.data.EduUserInfo;
import io.agora.education.api.user.data.EduUserLeftType;
import io.agora.education.api.user.data.EduUserRole;
import io.agora.education.api.user.data.EduUserStateChangeType;
import io.agora.edu.classroom.adapter.ClassVideoAdapter;
import io.agora.edu.common.bean.board.BoardState;
import io.agora.edu.classroom.bean.channel.Room;
import io.agora.edu.R2;
import io.agora.edu.classroom.fragment.UserListFragment;

import static io.agora.education.impl.Constants.AgoraLog;

public class SmallClassActivity extends BaseClassActivity implements TabLayout.OnTabSelectedListener {
    private static final String TAG = "SmallClassActivity";

    @BindView(R2.id.layout_placeholder)
    protected ConstraintLayout layout_placeholder;
    @BindView(R2.id.rcv_videos)
    protected RecyclerView rcv_videos;
    @BindView(R2.id.layout_im)
    protected View layout_im;
    @BindView(R2.id.layout_tab)
    protected TabLayout layout_tab;

    private ClassVideoAdapter classVideoAdapter;
    private UserListFragment userListFragment;
    private View teacherPlaceholderView;

    @Override
    protected int getLayoutResId() {
        return R.layout.activity_small_class;
    }

    @Override
    protected void initData() {
        super.initData();
        joinRoomAsStudent(getMainEduRoom(), agoraEduLaunchConfig.getUserName(), agoraEduLaunchConfig.getUserUuid(), true, true, true,
                new EduCallback<EduStudent>() {
                    @Override
                    public void onSuccess(@Nullable EduStudent res) {
                        runOnUiThread(() -> showFragmentWithJoinSuccess());
                        notifyStreamAndUser();
                        getMainEduRoom().getLocalUser(new EduCallback<EduUser>() {
                            @Override
                            public void onSuccess(@Nullable EduUser user) {
                                userListFragment.setLocalUserUuid(user.getUserInfo().getUserUuid());
                                initTitleTimeState();
                                initParseBoardInfo(getMainEduRoom());
                            }

                            @Override
                            public void onFailure(@NotNull EduError error) {
                            }
                        });
                    }

                    @Override
                    public void onFailure(@NotNull EduError error) {
                        joinFailed(error.getType(), error.getMsg());
                    }
                });
        classVideoAdapter = new ClassVideoAdapter();
    }

    @Override
    protected void initView() {
        super.initView();
        LinearLayoutManager layoutManager = new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false);
        rcv_videos.setLayoutManager(layoutManager);
        rcv_videos.addItemDecoration(new RecyclerView.ItemDecoration() {
            @Override
            public void getItemOffsets(@NonNull Rect outRect, @NonNull View view, @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
                super.getItemOffsets(outRect, view, parent, state);
                if (parent.getChildAdapterPosition(view) > 0) {
                    outRect.left = getResources().getDimensionPixelSize(R.dimen.dp_2_5);
                }
            }
        });
        rcv_videos.setAdapter(classVideoAdapter);
        layout_tab.addOnTabSelectedListener(this);
        userListFragment = new UserListFragment();
        getSupportFragmentManager().beginTransaction()
                .add(R.id.layout_chat_room, userListFragment)
                .show(userListFragment)
                .commitNow();
        findViewById(R.id.send1).setOnClickListener((v) -> {
        });
    }

    @Override
    protected int getClassType() {
        return Room.Type.SMALL;
    }

    @OnClick(R2.id.iv_float)
    public void onClick(View view) {
        boolean isSelected = view.isSelected();
        view.setSelected(!isSelected);
        layout_im.setVisibility(isSelected ? View.VISIBLE : View.GONE);
    }

    @Override
    public void onTabSelected(TabLayout.Tab tab) {
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        if (tab.getPosition() == 0) {
            transaction.show(chatRoomFragment).hide(userListFragment);
        } else {
            transaction.show(userListFragment).hide(chatRoomFragment);
        }
        transaction.commitNow();
    }

    @Override
    public void onTabUnselected(TabLayout.Tab tab) {

    }

    @Override
    public void onTabReselected(TabLayout.Tab tab) {

    }

    /**
     * 刷新流和user列表
     * 流：屏幕分享流和摄像头流
     */
    private void notifyStreamAndUser() {
        getCurFullStream(new EduCallback<List<EduStreamInfo>>() {
            @Override
            public void onSuccess(@Nullable List<EduStreamInfo> streams) {
                if (streams != null) {
                    for (EduStreamInfo streamInfo : streams) {
                        switch (streamInfo.getVideoSourceType()) {
                            case SCREEN:
                                runOnUiThread(() -> {
                                    layout_whiteboard.setVisibility(View.GONE);
                                    layout_share_video.setVisibility(View.VISIBLE);
                                    layout_share_video.removeAllViews();
                                    renderStream(getMainEduRoom(), streamInfo, layout_share_video);
                                });
                                break;
                            default:
                                break;
                        }
                    }
                    userListFragment.setUserList(streams);
                    showVideoList(streams);
                }
            }

            @Override
            public void onFailure(@NotNull EduError error) {

            }
        });
    }

    @Override
    public void onRemoteUsersInitialized(@NotNull List<? extends EduUserInfo> users, @NotNull EduRoom classRoom) {
        super.onRemoteUsersInitialized(users, classRoom);
    }

    @Override
    public void onRemoteUsersJoined(@NotNull List<? extends EduUserInfo> users, @NotNull EduRoom classRoom) {
        super.onRemoteUsersJoined(users, classRoom);
    }

    @Override
    public void onRemoteUserLeft(@NotNull EduUserEvent userEvent, @NotNull EduRoom classRoom) {
        super.onRemoteUserLeft(userEvent, classRoom);
    }

    @Override
    public void onRemoteUserUpdated(@NotNull EduUserEvent userEvent, @NotNull EduUserStateChangeType type,
                                    @NotNull EduRoom classRoom) {
        super.onRemoteUserUpdated(userEvent, type, classRoom);
    }

    @Override
    public void onRoomMessageReceived(@NotNull EduMsg message, @NotNull EduRoom classRoom) {
        super.onRoomMessageReceived(message, classRoom);
    }

    @Override
    public void onUserMessageReceived(@NotNull EduMsg message) {
        super.onUserMessageReceived(message);
    }

    @Override
    public void onRoomChatMessageReceived(@NotNull EduChatMsg eduChatMsg, @NotNull EduRoom classRoom) {
        super.onRoomChatMessageReceived(eduChatMsg, classRoom);
    }

    @Override
    public void onUserChatMessageReceived(@NotNull EduChatMsg chatMsg) {
        super.onUserChatMessageReceived(chatMsg);
    }

    @Override
    public void onRemoteStreamsInitialized(@NotNull List<? extends EduStreamInfo> streams, @NotNull EduRoom classRoom) {
        super.onRemoteStreamsInitialized(streams, classRoom);
        notifyStreamAndUser();
        classRoom.getLocalUser(new EduCallback<EduUser>() {
            @Override
            public void onSuccess(@Nullable EduUser user) {
                userListFragment.setLocalUserUuid(user.getUserInfo().getUserUuid());
            }

            @Override
            public void onFailure(@NotNull EduError error) {

            }
        });
    }

    @Override
    public void onRemoteStreamsAdded(@NotNull List<EduStreamEvent> streamEvents, @NotNull EduRoom classRoom) {
        super.onRemoteStreamsAdded(streamEvents, classRoom);
        boolean notify = false;
        for (EduStreamEvent streamEvent : streamEvents) {
            EduStreamInfo streamInfo = streamEvent.getModifiedStream();
            switch (streamInfo.getVideoSourceType()) {
                case CAMERA:
                    notify = true;
                    break;
                default:
                    break;
            }
        }
        if (notify) {
            AgoraLog.e(TAG + ":有远端Camera流添加，刷新视频列表");
        }
        notifyVideoUserList(notify);
    }

    @Override
    public void onRemoteStreamUpdated(@NotNull List<EduStreamEvent> streamEvents, @NotNull EduRoom classRoom) {
        super.onRemoteStreamUpdated(streamEvents, classRoom);
        boolean notify = false;
        for (EduStreamEvent streamEvent : streamEvents) {
            EduStreamInfo streamInfo = streamEvent.getModifiedStream();
            switch (streamInfo.getVideoSourceType()) {
                case CAMERA:
                    notify = true;
                    break;
                default:
                    break;
            }
        }
        if (notify) {
            AgoraLog.e(TAG, "有远端Camera流被修改，刷新视频列表");
        }
        notifyVideoUserList(notify);
    }

    @Override
    public void onRemoteStreamsRemoved(@NotNull List<EduStreamEvent> streamEvents, @NotNull EduRoom classRoom) {
        super.onRemoteStreamsRemoved(streamEvents, classRoom);
        boolean notify = false;
        for (EduStreamEvent streamEvent : streamEvents) {
            EduStreamInfo streamInfo = streamEvent.getModifiedStream();
            switch (streamInfo.getVideoSourceType()) {
                case CAMERA:
                    notify = true;
                    break;
                default:
                    break;
            }
        }
        if (notify) {
            AgoraLog.e(TAG + ":有远端Camera流被移除，刷新视频列表");
        }
        notifyVideoUserList(notify);
    }

    @Override
    public void onRoomStatusChanged(@NotNull EduRoomChangeType event, @NotNull EduUserInfo operatorUser, @NotNull EduRoom classRoom) {
        super.onRoomStatusChanged(event, operatorUser, classRoom);
    }

    @Override
    public void onRoomPropertiesChanged(@NotNull EduRoom classRoom, @Nullable Map<String, Object> cause) {
        super.onRoomPropertiesChanged(classRoom, cause);
    }

    @Override
    public void onNetworkQualityChanged(@NotNull NetworkQuality quality, @NotNull EduUserInfo user,
                                        @NotNull EduRoom classRoom) {
        super.onNetworkQualityChanged(quality, user, classRoom);
    }

    @Override
    public void onConnectionStateChanged(@NotNull ConnectionState state, @NotNull EduRoom classRoom) {
        super.onConnectionStateChanged(state, classRoom);
    }

    @Override
    public void onLocalUserUpdated(@NotNull EduUserEvent userEvent, @NotNull EduUserStateChangeType type) {
        super.onLocalUserUpdated(userEvent, type);
        /**更新视频列表和用户列表*/
        notifyVideoUserListForLocal();
    }

    @Override
    public void onLocalStreamAdded(@NotNull EduStreamEvent streamEvent) {
        super.onLocalStreamAdded(streamEvent);
        notifyVideoUserListForLocal();
    }

    @Override
    public void onLocalStreamUpdated(@NotNull EduStreamEvent streamEvent) {
        super.onLocalStreamUpdated(streamEvent);
        notifyVideoUserListForLocal();
    }

    @Override
    public void onLocalStreamRemoved(@NotNull EduStreamEvent streamEvent) {
        super.onLocalStreamRemoved(streamEvent);
        /**小班课场景下，此回调被调用就说明classroom结束，人员退出；所以此回调可以不处理*/
    }

    @Override
    public void onGlobalStateChanged(GlobalState state) {
        super.onGlobalStateChanged(state);
        BoardState boardState = (BoardState) state;
        List<String> grantedUuids = boardState.getGrantUsers();
        userListFragment.setGrantedUuids(grantedUuids);
    }

    private void showVideoList(List<EduStreamInfo> list) {
        runOnUiThread(() -> {
            for (int i = 0; i < list.size(); i++) {
                EduStreamInfo streamInfo = list.get(i);
                if (streamInfo.getPublisher().getRole().equals(EduUserRole.TEACHER)) {
                    /*隐藏老师的占位布局*/
                    layout_placeholder.setVisibility(View.GONE);
                    if (i != 0) {
                        Collections.swap(list, 0, i);
                    }
                    classVideoAdapter.setNewList(list);
                    return;
                }
            }
            /*显示老师的占位布局*/
            if (teacherPlaceholderView == null) {
                teacherPlaceholderView = LayoutInflater.from(this).inflate(R.layout.layout_video_small_class,
                        layout_placeholder);
            }
            layout_placeholder.setVisibility(View.VISIBLE);
            classVideoAdapter.setNewList(list);
        });
    }

    /**
     * 刷新视频列表和学生列表
     */
    private void notifyVideoUserList(boolean notifyCameraVideo) {
        getCurFullStream(new EduCallback<List<EduStreamInfo>>() {
            @Override
            public void onSuccess(@Nullable List<EduStreamInfo> streamInfos) {
                /**刷新视频列表*/
                if (notifyCameraVideo) {
                    showVideoList(streamInfos);
                }
                userListFragment.setUserList(streamInfos);
            }

            @Override
            public void onFailure(@NotNull EduError error) {

            }
        });
    }

    /**
     * 因为本地用户或本地流的改变而刷新视频列表和学生列表
     */
    private void notifyVideoUserListForLocal() {
        getCurFullStream(new EduCallback<List<EduStreamInfo>>() {
            @Override
            public void onSuccess(@Nullable List<EduStreamInfo> streamInfos) {
                showVideoList(streamInfos);
                userListFragment.updateLocalStream(getLocalCameraStream());
                userListFragment.setUserList(streamInfos);
            }

            @Override
            public void onFailure(@NotNull EduError error) {

            }
        });
    }

    @Override
    public void onLocalUserLeft(@NotNull EduUserEvent userEvent, @NotNull EduUserLeftType leftType) {

    }
}

File: edu/src/main/java/io/agora/edu/classroom/SmallClassActivity.kt
package io.agora.edu.classroom

import android.os.Bundle
import android.util.Log
import android.view.ViewTreeObserver.OnGlobalLayoutListener
import android.widget.RelativeLayout
import com.herewhite.sdk.domain.SDKError
import io.agora.edu.R
import com.herewhite.sdk.domain.SceneState
import io.agora.edu.launch.AgoraEduCourseware
import io.agora.edu.launch.AgoraEduSDK
import io.agora.education.api.base.EduError
import io.agora.education.api.room.EduRoom
import io.agora.education.api.user.EduStudent
import io.agora.education.api.message.EduChatMsg
import io.agora.education.api.room.data.EduRoomChangeType
import io.agora.education.api.stream.data.EduStreamEvent
import io.agora.education.api.stream.data.EduStreamInfo
import io.agora.education.api.user.data.*
import io.agora.educontext.EduContextError
import io.agora.educontext.WhiteboardDrawingConfig
import io.agora.rtc.IRtcEngineEventHandler
import io.agora.rtc.RtcChannel
import io.agora.uikit.impl.chat.tabs.ChatTabConfig
import io.agora.uikit.impl.chat.tabs.TabType
import io.agora.uikit.impl.container.AgoraContainerConfig
import io.agora.uikit.interfaces.protocols.AgoraUIContainer
import io.agora.uikit.impl.container.AgoraContainerType

class SmallClassActivity : BaseClassActivity() {
    private val tag = "SmallClassActivity"

    private val whiteBoardManagerEventListener = object : WhiteBoardManagerEventListener {
        override fun onWhiteBoardJoinSuccess(config: WhiteboardDrawingConfig) {
            getReporter().reportWhiteBoardResult("1", null, null)
            setWhiteboardJoinSuccess()
            checkProcessSuccess()
        }

        override fun onWhiteBoardJoinFail(error: SDKError?) {
            if (error != null) {
                container?.showError(EduContextError(-1, error.toString()))
            }
            getReporter().reportWhiteBoardResult("0", "White board join room fail", null)
        }

        override fun onSceneChanged(state: SceneState) {
            screenShareManager?.checkAndNotifyScreenShareByScene(state)
        }

        override fun onGrantedChanged() {
            userListManager?.notifyUserList()
        }
    }

    override var deviceManagerEventListener = object : DeviceManagerEventListener {
        override fun onCameraDeviceEnableChanged(enabled: Boolean) {
            userListManager?.updateLocalCameraSwitchState(!enabled)
            userListManager?.notifyUserList()
        }

        override fun onMicDeviceEnabledChanged(enabled: Boolean) {
            userListManager?.updateLocalMicSwitchState(!enabled)
            userListManager?.notifyUserList()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        contentLayout?.viewTreeObserver?.addOnGlobalLayoutListener(object : OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                if (contentLayout!!.width > 0 && contentLayout!!.height > 0) {
                    contentLayout!!.viewTreeObserver.removeOnGlobalLayoutListener(this)

                    if (EduDebugMode.useDebugUI) {
                        Log.i(tag, "create debug ui container")
                        container = AgoraUIContainer.create(
                                contentLayout!!,
                                0, 0,
                                contentLayout!!.width,
                                contentLayout!!.height,
                                AgoraContainerType.Debug,
                                eduContext,
                                AgoraContainerConfig(listOf()))
                    } else {
                        container = AgoraUIContainer.create(contentLayout!!,
                                0, 0, contentLayout!!.width,
                                contentLayout!!.height,
                                AgoraContainerType.SmallClass, eduContext,
                                AgoraContainerConfig(chatTabConfigs =
                                listOf(
                                        ChatTabConfig(getString(R.string.agora_chat_tab_message), TabType.Public, null),
                                        ChatTabConfig(getString(R.string.agora_chat_tab_private), TabType.Private, null)
                                )))
                    }

                    whiteboardContext.getHandlers()?.forEach {
                        it.getBoardContainer()?.let { viewGroup ->
                            whiteBoardContainer = viewGroup
                            whiteBoardManager = WhiteBoardManager(this@SmallClassActivity,
                                    launchConfig!!, viewGroup, whiteboardContext)
                            return@forEach
                        }
                    }

                    whiteBoardManager?.let {
                        launchConfig?.let { config ->
                            var ware: AgoraEduCourseware? = null
                            if (AgoraEduSDK.COURSEWARES.size > 0) {
                                ware = AgoraEduSDK.COURSEWARES[0]
                            }
                            it.initData(config.roomUuid, config.whiteBoardAppId,
                                    preCheckData?.board?.boardRegion, ware)
                        }
                        it.whiteBoardManagerEventListener = whiteBoardManagerEventListener
                    }
                }
            }
        })
    }

    override fun onContentViewLayout(): RelativeLayout {
        contentLayout = RelativeLayout(this)
        return contentLayout!!
    }

    override fun onRoomJoinConfig(): JoinRoomConfiguration {
        return JoinRoomConfiguration(
                autoPublish = false,
                autoSubscribe = true,
                needUserListener = true)
    }

    override fun onRoomJoined(success: Boolean, student: EduStudent?, error: EduError?) {
        super.onRoomJoined(success, student, error)
    }

    override fun onDestroy() {
        super.onDestroy()
        teacherVideoManager?.dispose()
        screenShareManager?.dispose()
        userListManager?.dispose()
        handsUpManager?.dispose()
        roomStateManager?.dispose()
    }

    override fun onRemoteUsersInitialized(users: List<EduUserInfo>, classRoom: EduRoom) {
        super.onRemoteUsersInitialized(users, classRoom)
    }

    override fun onRemoteUsersJoined(users: List<EduUserInfo>, classRoom: EduRoom) {
        super.onRemoteUsersJoined(users, classRoom)
        userListManager?.notifyUserList()
    }

    override fun onRemoteUserUpdated(userEvent: EduUserEvent, type: EduUserStateChangeType, classRoom: EduRoom) {
        super.onRemoteUserUpdated(userEvent, type, classRoom)
    }

    override fun onRemoteUserLeft(userEvent: EduUserEvent, classRoom: EduRoom) {
        super.onRemoteUserLeft(userEvent, classRoom)
        userListManager?.notifyUserList()
    }

    override fun onRoomChatMessageReceived(chatMsg: EduChatMsg, classRoom: EduRoom) {
        super.onRoomChatMessageReceived(chatMsg, classRoom)
        chatManager?.receiveRemoteChatMessage(chatMsg)
    }

    override fun onRemoteStreamsInitialized(streams: List<EduStreamInfo>, classRoom: EduRoom) {
        super.onRemoteStreamsInitialized(streams, classRoom)
        screenShareManager?.checkAndNotifyScreenShareRestored()
    }

    override fun onRemoteStreamsAdded(streamEvents: MutableList<EduStreamEvent>, classRoom: EduRoom) {
        super.onRemoteStreamsAdded(streamEvents, classRoom)
        teacherVideoManager?.notifyUserDetailInfo(EduUserRole.TEACHER)
        screenShareManager?.checkAndNotifyScreenShareStarted(streamEvents)
        if (streamEvents.find { it.modifiedStream.publisher.role == EduUserRole.TEACHER } == null) {
            userListManager?.notifyUserList()
        }
    }

    override fun onRemoteStreamUpdated(streamEvents: MutableList<EduStreamEvent>, classRoom: EduRoom) {
        super.onRemoteStreamUpdated(streamEvents, classRoom)
        teacherVideoManager?.notifyUserDetailInfo(EduUserRole.TEACHER)
        screenShareManager?.checkAndNotifyScreenShareStarted(streamEvents)
        if (streamEvents.find { it.modifiedStream.publisher.role == EduUserRole.TEACHER } == null) {
            userListManager?.notifyUserList()
        }
    }

    override fun onRemoteStreamsRemoved(streamEvents: MutableList<EduStreamEvent>, classRoom: EduRoom) {
        super.onRemoteStreamsRemoved(streamEvents, classRoom)
        teacherVideoManager?.notifyUserDetailInfo(EduUserRole.TEACHER)
        screenShareManager?.checkAndNotifyScreenShareRemoved(streamEvents)
        if (streamEvents.find { it.modifiedStream.publisher.role == EduUserRole.TEACHER } == null) {
            userListManager?.notifyUserList()
        }
    }

    override fun onLocalStreamAdded(streamEvent: EduStreamEvent) {
        super.onLocalStreamAdded(streamEvent)
        userListManager?.addLocalStream(streamEvent)
    }

    override fun onLocalStreamUpdated(streamEvent: EduStreamEvent) {
        super.onLocalStreamUpdated(streamEvent)
        userListManager?.updateLocalStream(streamEvent)
        userListManager?.notifyUserList()
    }

    override fun onLocalStreamRemoved(streamEvent: EduStreamEvent) {
        super.onLocalStreamRemoved(streamEvent)
        userListManager?.removeLocalStream(streamEvent)
        userListManager?.notifyUserList()
    }

    override fun onRoomStatusChanged(type: EduRoomChangeType, operatorUser: EduUserInfo?, classRoom: EduRoom) {
        super.onRoomStatusChanged(type, operatorUser, classRoom)
        roomStateManager?.updateClassState(type)
    }

    override fun onRoomPropertiesChanged(changedProperties: MutableMap<String, Any>,
                                         classRoom: EduRoom, cause: MutableMap<String, Any>?,
                                         operator: EduBaseUserInfo?) {
        super.onRoomPropertiesChanged(changedProperties, classRoom, cause, operator)
        handsUpManager?.notifyHandsUpEnable(cause)
        handsUpManager?.notifyHandsUpState(cause)
        userListManager?.notifyListByPropertiesChanged(cause)
    }

    override fun onRemoteVideoStateChanged(rtcChannel: RtcChannel?, uid: Int, state: Int, reason: Int, elapsed: Int) {
        super.onRemoteVideoStateChanged(rtcChannel, uid, state, reason, elapsed)
        teacherVideoManager?.notifyUserDetailInfo(EduUserRole.TEACHER)
    }

    override fun onLocalVideoStateChanged(localVideoState: Int, error: Int) {
        super.onLocalVideoStateChanged(localVideoState, error)
        userListManager?.updateLocalCameraAvailableState(localVideoState)
    }

    override fun onLocalAudioStateChanged(localAudioState: Int, error: Int) {
        super.onLocalAudioStateChanged(localAudioState, error)
        userListManager?.updateLocalMicAvailableState(localAudioState)
    }

    override fun onAudioVolumeIndicationOfLocalSpeaker(speakers: Array<out IRtcEngineEventHandler.AudioVolumeInfo>?, totalVolume: Int) {
        super.onAudioVolumeIndicationOfLocalSpeaker(speakers, totalVolume)
        teacherVideoManager?.updateAudioVolume(speakers)
        userListManager?.updateAudioVolumeIndication(speakers)

    }

    override fun onAudioVolumeIndicationOfRemoteSpeaker(speakers: Array<out IRtcEngineEventHandler.AudioVolumeInfo>?, totalVolume: Int) {
        super.onAudioVolumeIndicationOfRemoteSpeaker(speakers, totalVolume)
        teacherVideoManager?.updateAudioVolume(speakers)
        userListManager?.updateAudioVolumeIndication(speakers)
    }

    override fun onRemoteUserPropertiesChanged(changedProperties: MutableMap<String, Any>,
                                               classRoom: EduRoom, userInfo: EduUserInfo,
                                               cause: MutableMap<String, Any>?, operator: EduBaseUserInfo?) {
        super.onRemoteUserPropertiesChanged(changedProperties, classRoom, userInfo, cause, operator)
        teacherVideoManager?.updateRemoteDeviceState(userInfo, cause)
        teacherVideoManager?.notifyUserDetailInfo(EduUserRole.TEACHER)
        userListManager?.updateRemoteDeviceState(userInfo, cause)
        userListManager?.notifyListByPropertiesChanged(cause)
        chatManager?.notifyUserChatMuteStatus(userInfo, cause, operator)
    }

    override fun onLocalUserPropertiesChanged(changedProperties: MutableMap<String, Any>,
                                              userInfo: EduUserInfo, cause: MutableMap<String, Any>?,
                                              operator: EduBaseUserInfo?) {
        super.onLocalUserPropertiesChanged(changedProperties, userInfo, cause, operator)
        userListManager?.notifyListByPropertiesChanged(cause)
        chatManager?.notifyUserChatMuteStatus(userInfo, cause, operator)
    }
}